x-pack/platform/plugins/shared/observability_ai_assistant/server/service/index_assets/create_or_update_knowledge_base_index_assets.ts

```diff
@@ -0,0 +1,102 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { CoreSetup, Logger } from '@kbn/core/server';
+import type { ObservabilityAIAssistantPluginStartDependencies } from '../../types';
+import {
+  DEFAULT_INFERENCE_ENDPOINT,
+  getComponentTemplate,
+} from './templates/kb_component_template';
+import { resourceNames } from '..';
+import { createKnowledgeBaseIndex } from '../knowledge_base_service/create_knowledge_base_index';
+import { updateKnowledgeBaseWriteIndexAlias } from '../knowledge_base_service/update_knowledge_base_index_alias';
+import { hasKbIndex } from '../knowledge_base_service/has_kb_index';
+import { updateKnowledgeBaseIndexMapping } from '../knowledge_base_service/update_knowledge_base_index_mapping';
+
+export async function createOrUpdateKnowledgeBaseIndexAssets({
+  logger,
+  core,
+  inferenceId = DEFAULT_INFERENCE_ENDPOINT, // TODO: use `.elser-v2-elastic` for serverless on EIS
+}: {
+  logger: Logger;
+  core: CoreSetup<ObservabilityAIAssistantPluginStartDependencies>;
+  inferenceId?: string;
+}) {
+  try {
+    logger.debug('Setting up knowledge base index assets');
+    const [coreStart] = await core.getStartServices();
+    const { asInternalUser } = coreStart.elasticsearch.client;
+
+    // Knowledge base: component template
+    await asInternalUser.cluster.putComponentTemplate({
+      create: false,
+      name: resourceNames.componentTemplate.kb,
+      template: getComponentTemplate(inferenceId),
+    });
+
+    // Knowledge base: index template
+    await asInternalUser.indices.putIndexTemplate({
+      name: resourceNames.indexTemplate.kb,
+      composed_of: [resourceNames.componentTemplate.kb],
+      create: false,
+      index_patterns: [resourceNames.indexPatterns.kb],
+      template: {
+        settings: {
+          number_of_shards: 1,
+          auto_expand_replicas: '0-1',
+          hidden: true,
+        },
+      },
+    });
+
+    const indexExists = await hasKbIndex({ esClient: coreStart.elasticsearch.client });
+
+    // Knowledge base: update write index if index exists
+    if (indexExists) {
+      const kbWriteIndexAlias = resourceNames.writeIndexAlias.kb;
+
+      const aliasInfo = await asInternalUser.indices.getAlias({
+        name: kbWriteIndexAlias,
+        ignore_unavailable: true,
+      });
+
+      const isWriteIndex = Object.entries(aliasInfo).some(
+        ([_indexName, aliasData]) => aliasData.aliases?.[kbWriteIndexAlias]?.is_write_index
+      );
+
+      if (isWriteIndex) {
+        logger.info(
+          `Knowledge base write index [${kbWriteIndexAlias}] already exists and is assigned as a write index. Skipping creation.`
+        );
+
+        // Knowledge base: update write index
+        updateKnowledgeBaseIndexMapping({ esClient: coreStart.elasticsearch.client, logger });
+      }
+
+      return;
+    }
```

Comment:
I don't think we need to check whether it is the write index. If the `writeIndexAlias` is pointing to it, we can assume it is the write index imo.

```suggestion
    // Knowledge base: update write index if index exists
    const indexExists = await hasKbIndex({ esClient: coreStart.elasticsearch.client });
    if (indexExists) {
      await updateKnowledgeBaseIndexMapping({ esClient: coreStart.elasticsearch.client, logger });
      return;
    }
```